### PR TITLE
[Gekidou MM-47655] Implement Add Members screen - (merge to main)

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This PR implements the `Add Members` Screen.  It is available by clicking the `Add People` box which can be located via the following ways:

 1) Channel > `...` in header 
 2) Channel > tap the channel header

 This PR does not implement the bottom selected users panel at this time. It is in a separate PR. https://github.com/mattermost/mattermost-mobile/pull/6699

Todo: 
- [ ] guest cannot add people, do we hide the add people box from them?
- [x] return to the Channel Info screen after adding Members
- [x] add toast after adding members
- [x] after removing a user, update the members list to reflect the removed user
- [x] update profile bottom sheet - ticket created https://mattermost.atlassian.net/browse/MM-48470
- [x] fix warning that occurs after selecting a user profile
- [x] needs 'no more users to add to the channel' empty results page
- [x] tutorial should highlight the first row (verify in tablet and phone).  Matt says this was working before so might be introducing a regression.
- [ ] need to pop to correct screen in both cases where add people box can be accessed (look at `Start Call` box for example)
  - [ ] `Channel Info` > `Add People` 
  - [ ] `Channel Nav Header` > `...` > `Add People`
- [ ] `Go Back` button when no members left to add also needs to know which screen to `go back` to

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47655

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note

```

### CODE REVIEW

File: app/actions/remote/user.ts
Comment (enahum) [inline app/actions/remote/user.ts:552]: same as in the other PR

File: app/screens/channel_add_people/channel_add_people.tsx
Comment (enahum) [inline app/screens/channel_add_people/channel_add_people.tsx:308]: ???

File: app/screens/channel_add_people/channel_add_people.tsx
Comment (enahum) [inline app/screens/channel_add_people/channel_add_people.tsx:9]: ??

File: app/screens/channel_add_people/channel_add_people.tsx
Comment (enahum) [inline app/screens/channel_add_people/channel_add_people.tsx:12]:
```suggestion
import useDidUpdate from '@hooks/did_update';
```

**Review by enahum — DISMISSED** [review]

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
@jfrerich a few things from me:

1. Since we're not implementing this with the selected users bottom panel for now (as you described in the PR description), we should use the chips as we do in the new DM modal for now until the bottom panel is implemented. Otherwise you won't be able to see who you've selected as you scroll through a long list
2. As I type in the search input, it immediately shows an empty state at first and then the loader displays. Can we avoid displaying the empty state until we're certain there are no results? This may apply to other search scenarios in the app as well. I feel like I've called this out before.
3. In a channel where all team members are already in the channel, there are no members to add, so I guess we'll need a slightly different empty state to show here. In this case we should probably hide the search input as well.

image

We can just do something like this ([figma link here](https://www.figma.com/file/uCS4dYCHSS5NO7PjlVUBzI/MM-19947-Mobile-Multi-select?node-id=3154%3A96661&t=XvFpID6yqJVD8O8I-4) as well): 




4. In the design, a toast was supposed to display when you go back to the main channel info screen


5. If we can go the toast approach, I'm hoping we won't need this loading state (see below)? If we do still need this loading state, I would center the loader on the screen instead of where it is positioned right now.
image

6. I can't seem to consistently reproduce this, but there were a few times during my testing where the `>` back button disappeared from the modal's top bar so I couldn't go back to the info screen. In this scenario, if I hit the `Add` button, the screen hung up on the loading state and never finished and 'Possible unhandled promise' error displayed in the simulator. I had to reload the entire app to get out of this state.

image

image

7. On the channel info screen, tapping 'Members' resulted in this 'Add members' screen. This should not be the case. I know you're implementing the view/manage members screen separately though, so this may not be an issue.

**Comment by jfrerich** [comment]
Here's an update on your response, @matthewbirtch 

> 1. Since we're not implementing this with the selected users bottom panel for now (as you described in the PR description), we should use the chips as we do in the new DM modal for now until the bottom panel is implemented. Otherwise you won't be able to see who you've selected as you scroll through a long list

**SUGGESTION** The issue with the selected users and why it probably wasn't used in V1 is because it isn't scrollable.  I propose having it added after https://github.com/mattermost/mattermost-mobile/pull/6699 is merged.  I have added a ticket to do this after merging that PR

> 2. As I type in the search input, it immediately shows an empty state at first and then the loader displays. Can we avoid displaying the empty state until we're certain there are no results? This may apply to other search scenarios in the app as well. I feel like I've called this out before.

**FIXED** for `AddMembers` and `CreateDirectMessage` screens.  The loading spinner shows instead of the emtpy state.

> 3. In a channel where all team members are already in the channel, there are no members to add, so I guess we'll need a slightly different empty state to show here. In this case we should probably hide the search input as well.

**FIXED** I added the screen you designed. with the back button.

> 4. In the design, a toast was supposed to display when you go back to the main channel info screen

**IN PROGRESS**

> 5. If we can go the toast approach, I'm hoping we won't need this loading state (see below)? If we do still need this loading state, I would center the loader on the screen instead of where it is positioned right now.

**IN PROGRESS**

> 6. I can't seem to consistently reproduce this, but there were a few times during my testing where the > back button disappeared from the modal's top bar so I couldn't go back to the info screen. In this scenario, if I hit the Add button, the screen hung up on the loading state and never finished and 'Possible unhandled promise' error displayed in the simulator. I had to reload the entire app to get out of this state.

**IN PROGRESS** I am seeing this but don't have a clear set of repro steps yet.  I'll continue working on this one. 

> 7. On the channel info screen, tapping 'Members' resulted in this 'Add members' screen. This should not be the case. I know you're implementing the view/manage members screen separately though, so this may not be an issue.

**CORRECT** that screen will be implemented via https://github.com/mattermost/mattermost-mobile/pull/6771

File: assets/base/i18n/en.json
Comment (jfrerich) [inline assets/base/i18n/en.json:862]: use intl `plural` option and remove second id  https://formatjs.io/docs/intl-messageformat/

File: app/screens/channel_add_people/channel_add_people.tsx
Comment (jfrerich) [reply app/screens/channel_add_people/channel_add_people.tsx]: selected users panel added and now being use in this PR.

**Review by avinashlng1080 — APPROVED** [review]
LGTM. Approving so as not to blocked PR.  However, some lines are duplicated over 1 or 2 files.  Can these duplicates be reduced?
code duplicates

**Comment by avinashlng1080** [comment]
@jfrerich to resolve merge conflict

File: app/screens/channel_add_people/index.tsx
Comment (jfrerich) [reply app/screens/channel_add_people/index.tsx]: The need the currentChannel observable for `isGroupConstrained` and to get `channelId`.  Thats why `currentChannel` is it's own variable

**Comment by Willyfrog** [comment]
what's left on this PR to be merged?

**Comment by esethna** [comment]
@amyblais we discussed potentially getting an exception to take the members related PRs into v2.1 given the impact on user experience and feedback in the stores. The risk can be mitigated by getting it on the beta app launch tomorrow and testing there thoroughly. 

If we're in alignment let's aim to cherry pick these into v2.1

**Comment by matthewbirtch** [comment]
@jfrerich we still need to add the 'plus' button for adding people on the View/Manage Members screen, right? Or is that going to be done in another PR?

image

**Comment by jfrerich** [comment]
@matthewbirtch, I've noticed those changes too. This was after merging latest master. I'm making updates now.

**Comment by jfrerich** [comment]
> what's left on this PR to be merged?

@Willyfrog, I'm currently reviewing changes need after we moved the new react-native bottom sheet.  See @matthewbirtch comment above. 

I have a fix for his comment and am working on the tablet version fixes.  tapping on search in tablet is wrong after the merge.

**Comment by enahum** [comment]
closing this PR for now

**Pull request closed without merging**
